File: packages/sparkling-method/android/src/main/java/com/tiktok/sparkling/method/registry/api/SparklingMethodInvocationCenter.kt
// Copyright (c) 2026 TikTok Pte. Ltd.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

package com.tiktok.sparkling.method.registry.api

import com.tiktok.sparkling.method.protocol.entity.BridgeCall
import com.tiktok.sparkling.method.registry.core.BridgePlatformType
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Central hub the Sparkling Bridge implementation uses to broadcast every
 * method invocation lifecycle event so debug tools (or analytics) can observe
 * them without coupling directly to the bridge implementation.
 *
 * The hub is intentionally minimal: payloads are passed as-is, callers are
 * expected to deep-copy or stringify on the listener side if needed.
 */
object SparklingMethodInvocationCenter {
    /**
     * Snapshot describing one method invocation. `id` correlates [onStart]
     * with the matching [onEnd]. Most fields are nullable because earlier
     * lifecycle events may not know them yet.
     */
    data class Event(
        val id: String,
        val name: String,
        val namespace: String?,
        val platform: BridgePlatformType,
        val params: Any?,
        val result: Any?,
        val code: Int?,
        val startTimeMs: Long,
        val endTimeMs: Long?,
    )

    interface Observer {
        /** Invoked on the bridge dispatcher thread before the method runs. */
        fun onStart(event: Event) = Unit

        /** Invoked once the method has produced its final result. */
        fun onEnd(event: Event)
    }

    private val observers = CopyOnWriteArrayList<Observer>()

    @JvmStatic
    fun addObserver(observer: Observer) {
        if (!observers.contains(observer)) observers.add(observer)
    }

    @JvmStatic
    fun removeObserver(observer: Observer) {
        observers.remove(observer)
    }

    @JvmStatic
    fun hasObservers(): Boolean = observers.isNotEmpty()

    /**
     * Notify observers that an invocation has started. The implementation
     * is fire-and-forget: any observer exception is swallowed so the bridge
     * remains unaffected.
     */
    @JvmStatic
    fun notifyStart(
        id: String,
        call: BridgeCall,
        namespace: String?,
        platform: BridgePlatformType,
        startTimeMs: Long,
    ) {
        if (observers.isEmpty()) return
        val event =
            Event(
                id = id,
                name = call.bridgeName,
                namespace = namespace,
                platform = platform,
                params = call.params,
                result = null,
                code = null,
                startTimeMs = startTimeMs,
                endTimeMs = null,
            )
        observers.forEach { runCatching { it.onStart(event) } }
    }

    @JvmStatic
    fun notifyEnd(
        id: String,
        call: BridgeCall,
        namespace: String?,
        platform: BridgePlatformType,
        startTimeMs: Long,
        endTimeMs: Long,
        result: Any?,
        code: Int?,
    ) {
        if (observers.isEmpty()) return
        val event =
            Event(
                id = id,
                name = call.bridgeName,
                namespace = namespace,
                platform = platform,
                params = call.params,
                result = result,
                code = code,
                startTimeMs = startTimeMs,
                endTimeMs = endTimeMs,
            )
        observers.forEach { runCatching { it.onEnd(event) } }
    }

    @JvmStatic
    fun notifyNativeToJsEvent(
        name: String,
        params: Any?,
        containerId: String?,
    ) {
        if (observers.isEmpty()) return
        val now = System.currentTimeMillis()
        val event =
            Event(
                id =
                    java.util.UUID
                        .randomUUID()
                        .toString(),
                name = name,
                namespace = "native-to-js",
                platform = BridgePlatformType.LYNX,
                params =
                    mapOf(
                        "event" to name,
                        "params" to params,
                        "containerID" to containerId,
                    ),
                result =
                    mapOf(
                        "status" to "sent",
                        "direction" to "Native -> JS",
                    ),
                code = 1,
                startTimeMs = now,
                endTimeMs = now,
            )
        observers.forEach { runCatching { it.onEnd(event) } }
    }
}

File: packages/sparkling-method/ios/Sources/Core/Pipe/SparklingMethodInvocationCenter.swift
// Copyright (c) 2026 TikTok Pte. Ltd.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

import Foundation

/// Snapshot describing one Sparkling Method invocation. The fields mirror the
/// Android `SparklingMethodInvocationCenter.Event` so debug surfaces can be
/// kept consistent across platforms.
public struct SparklingMethodInvocationEvent {
    public let id: String
    public let name: String
    public let namespace: String?
    public let platform: String
    public let params: [String: Any]?
    public let result: [String: Any]?
    public let statusCode: Int?
    public let statusMessage: String?
    public let startTime: Date
    public let endTime: Date?

    public var durationMs: Int? {
        guard let end = endTime else { return nil }
        return Int((end.timeIntervalSince(startTime)) * 1000)
    }
}

/// Observer protocol for [SparklingMethodInvocationCenter].
public protocol SparklingMethodInvocationObserver: AnyObject {
    func sparklingMethodInvocationDidStart(_ event: SparklingMethodInvocationEvent)
    func sparklingMethodInvocationDidEnd(_ event: SparklingMethodInvocationEvent)
}

public extension SparklingMethodInvocationObserver {
    func sparklingMethodInvocationDidStart(_ event: SparklingMethodInvocationEvent) {}
}

/// Central hub the Sparkling Method pipe uses to broadcast every invocation
/// lifecycle event. Debug tools attach themselves at app startup; the pipe
/// only walks observers when at least one is registered.
public final class SparklingMethodInvocationCenter {
    public static let shared = SparklingMethodInvocationCenter()

    /// Notification posted whenever a method invocation starts. The
    /// `Notification.object` is the `SparklingMethodInvocationCenter` shared
    /// instance and the underlying `SparklingMethodInvocationEvent` is
    /// available under `userInfo[SparklingMethodInvocationCenter.eventKey]`.
    /// Debug surfaces that don't want to take a hard dependency on
    /// `SparklingMethod` can subscribe by name string instead of importing
    /// this module.
    public static let didStartNotification = Notification.Name("SparklingMethodInvocationCenter.didStart")
    public static let didEndNotification = Notification.Name("SparklingMethodInvocationCenter.didEnd")
    public static let eventKey = "event"

    private let lock = NSLock()
    private var observers = NSHashTable<AnyObject>.weakObjects()

    private init() {}

    public func addObserver(_ observer: SparklingMethodInvocationObserver) {
        lock.lock(); defer { lock.unlock() }
        observers.add(observer as AnyObject)
    }

    public func removeObserver(_ observer: SparklingMethodInvocationObserver) {
        lock.lock(); defer { lock.unlock() }
        observers.remove(observer as AnyObject)
    }

    public var hasObservers: Bool {
        lock.lock(); defer { lock.unlock() }
        return observers.allObjects.contains { _ in true }
    }

    func notifyStart(_ event: SparklingMethodInvocationEvent) {
        for case let observer as SparklingMethodInvocationObserver in snapshot() {
            observer.sparklingMethodInvocationDidStart(event)
        }
        NotificationCenter.default.post(
            name: SparklingMethodInvocationCenter.didStartNotification,
            object: self,
            userInfo: makeNotificationUserInfo(for: event)
        )
    }

    func notifyEnd(_ event: SparklingMethodInvocationEvent) {
        for case let observer as SparklingMethodInvocationObserver in snapshot() {
            observer.sparklingMethodInvocationDidEnd(event)
        }
        NotificationCenter.default.post(
            name: SparklingMethodInvocationCenter.didEndNotification,
            object: self,
            userInfo: makeNotificationUserInfo(for: event)
        )
    }

    func notifyNativeToJsEvent(name: String, params: [String: Any]?, containerID: String?) {
        let now = Date()
        let payload: [String: Any] = [
            "event": name,
            "params": params ?? [:],
            "containerID": containerID ?? NSNull(),
        ]
        let result: [String: Any] = [
            "status": "sent",
            "direction": "Native -> JS",
        ]
        notifyEnd(
            SparklingMethodInvocationEvent(
                id: UUID().uuidString,
                name: name,
                namespace: "native-to-js",
                platform: "lynx",
                params: payload,
                result: result,
                statusCode: 1,
                statusMessage: "sent",
                startTime: now,
                endTime: now
            )
        )
    }

    func notifyCompletedInvocation(
        name: String,
        namespace: String?,
        platform: String,
        params: [String: Any]?,
        result: [String: Any]?,
        statusCode: Int?,
        statusMessage: String?
    ) {
        let now = Date()
        notifyEnd(
            SparklingMethodInvocationEvent(
                id: UUID().uuidString,
                name: name,
                namespace: namespace,
                platform: platform,
                params: params,
                result: result,
                statusCode: statusCode,
                statusMessage: statusMessage,
                startTime: now,
                endTime: now
            )
        )
    }

    /// Build a stable, type-erased dictionary representation of the event so
    /// debug surfaces can subscribe via `Notification.Name` strings without
    /// importing this module. Keys mirror the struct field names.
    private func makeNotificationUserInfo(for event: SparklingMethodInvocationEvent) -> [AnyHashable: Any] {
        var info: [AnyHashable: Any] = [
            SparklingMethodInvocationCenter.eventKey: event,
            "id": event.id,
            "name": event.name,
            "platform": event.platform,
            "startTime": event.startTime,
        ]
        if let namespace = event.namespace { info["namespace"] = namespace }
        if let params = event.params { info["params"] = params }
        if let result = event.result { info["result"] = result }
        if let statusCode = event.statusCode { info["statusCode"] = statusCode }
        if let statusMessage = event.statusMessage { info["statusMessage"] = statusMessage }
        if let endTime = event.endTime { info["endTime"] = endTime }
        return info
    }

    private func snapshot() -> [AnyObject] {
        lock.lock(); defer { lock.unlock() }
        return observers.allObjects
    }
}
